src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseService.java

```diff
@@ -695,6 +695,42 @@ public void combineAllCommitsOfRepositoryIntoOne(VcsRepositoryUrl repoUrl) throw
         gitService.combineAllCommitsIntoInitialCommit(exerciseRepository);
     }
 
+    /**
+     * Updates the timeline attributes of the given programming exercise
+     * @param updatedProgrammingExercise containing the changes that have to be saved
+     * @param notificationText optional text for a notification to all students about the update
+     * @return the updated ProgrammingExercise object.
+     * @throws EntityNotFoundException if there is no ProgrammingExercise for the given id.
+     * @throws IllegalAccessException  if the user does not have permissions to access the ProgrammingExercise.
+     */
+    public ProgrammingExercise updateTimeline(ProgrammingExercise updatedProgrammingExercise, @Nullable String notificationText)
+            throws EntityNotFoundException, IllegalAccessException {
+
+        Optional<ProgrammingExercise> programmingExercise = programmingExerciseRepository.findById(updatedProgrammingExercise.getId());
+        if (programmingExercise.isPresent()) {
+            programmingExercise.get().setReleaseDate(updatedProgrammingExercise.getReleaseDate());
+            programmingExercise.get().setDueDate(updatedProgrammingExercise.getDueDate());
+            programmingExercise.get().setBuildAndTestStudentSubmissionsAfterDueDate(updatedProgrammingExercise.getBuildAndTestStudentSubmissionsAfterDueDate());
+            programmingExercise.get().setAssessmentDueDate(updatedProgrammingExercise.getAssessmentDueDate());
+        }
+        else {
+            throw new EntityNotFoundException("Programming exercise not found with id: " + updatedProgrammingExercise.getId());
+        }
+
+        User user = userService.getUserWithGroupsAndAuthorities();
+        Course course = updatedProgrammingExercise.getCourseViaExerciseGroupOrCourseMember();
```

Comment:
I think you should check this earlier

Reply:
Moved to ProgrammingExerciseResource.java with 
35a4022